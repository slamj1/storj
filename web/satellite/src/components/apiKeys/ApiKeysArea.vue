// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="api-keys-area">
        <h1 class="api-keys-area__title" v-if="!isEmpty">API Keys</h1>
        <div class="api-keys-area__container">
            <ApiKeysCreationPopup
                @closePopup="closeNewApiKeyPopup"
                @showCopyPopup="showCopyApiKeyPopup"
                :is-popup-shown="isNewApiKeyPopupShown"
            />
            <ApiKeysCopyPopup
                :is-popup-shown="isCopyApiKeyPopupShown"
                :api-key-secret="apiKeySecret"
                @closePopup="closeCopyNewApiKeyPopup"
            />
            <div v-if="isHeaderShown" class="api-keys-header">
                <VHeader
                    ref="headerComponent"
                    placeholder="API Key"
                    :search="onSearchQueryCallback">
                    <div class="header-default-state" v-if="isDefaultHeaderState">
                        <VButton
                            class="button"
                            label="+ Create API Key"
                            width="180px"
                            height="48px"
                            :on-press="onCreateApiKeyClick"
                        />
                    </div>
                    <div class="header-selected-api-keys" v-if="areApiKeysSelected">
                        <VButton
                            class="button deletion"
                            label="Delete"
                            width="122px"
                            height="48px"
                            :on-press="onFirstDeleteClick"
                        />
                        <VButton
                            class="button"
                            label="Cancel"
                            width="122px"
                            height="48px"
                            is-white="true"
                            :on-press="onClearSelection"
                        />
                        <span class="header-selected-api-keys__info-text"><b>{{selectedAPIKeysCount}}</b> API Keys selected</span>
                    </div>
                    <div class="header-after-delete-click" v-if="areSelectedApiKeysBeingDeleted">
                        <span class="header-after-delete-click__confirmation-label">Are you sure you want to delete <b>{{selectedAPIKeysCount}}</b> {{apiKeyCountTitle}} ?</span>
                        <div class="header-after-delete-click__button-area">
                            <VButton
                                class="button deletion"
                                label="Delete"
                                width="122px"
                                height="48px"
                                :on-press="onDelete"
                            />
                            <VButton
                                class="button"
                                label="Cancel"
                                width="122px"
                                height="48px"
                                is-white="true"
                                :on-press="onClearSelection"
                            />
                        </div>
                    </div>
                </VHeader>
                <div class="blur-content" v-if="isDeleteClicked"></div>
                <div class="blur-search" v-if="isDeleteClicked"></div>
            </div>
            <div v-if="!isEmpty" class="api-keys-items">
                <SortingHeader :on-header-click-callback="onHeaderSectionClickCallback"/>
                <div class="api-keys-items__content">
                    <VList
                        :data-set="apiKeyList"
                        :item-component="itemComponent"
                        :on-item-click="toggleSelection"
                    />
                </div>
                <VPagination
                    v-if="totalPageCount > 1"
                    class="pagination-area"
                    ref="pagination"
                    :total-page-count="totalPageCount"
                    :on-page-click-callback="onPageClick"
                />
            </div>
            <div class="empty-search-result-area" v-if="isEmptySearchResultShown">
                <h1 class="empty-search-result-area__title">No results found</h1>
                <EmptySearchResultIcon class="empty-search-result-area__image"/>
            </div>
            <NoApiKeysArea
                :on-button-click="onCreateApiKeyClick"
                v-if="isEmptyStateShown"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import ApiKeysItem from '@/components/apiKeys/ApiKeysItem.vue';
import NoApiKeysArea from '@/components/apiKeys/NoApiKeysArea.vue';
import SortingHeader from '@/components/apiKeys/SortingHeader.vue';
import VButton from '@/components/common/VButton.vue';
import VHeader from '@/components/common/VHeader.vue';
import VList from '@/components/common/VList.vue';
import VPagination from '@/components/common/VPagination.vue';

import EmptySearchResultIcon from '@/../static/images/common/emptySearchResult.svg';

import { ApiKey, ApiKeyOrderBy } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';
import { API_KEYS_ACTIONS } from '@/utils/constants/actionNames';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';

import ApiKeysCopyPopup from './ApiKeysCopyPopup.vue';
import ApiKeysCreationPopup from './ApiKeysCreationPopup.vue';

// header state depends on api key selection state
enum HeaderState {
    DEFAULT = 0,
    ON_SELECT,
}

const {
    FETCH,
    DELETE,
    TOGGLE_SELECTION,
    CLEAR,
    CLEAR_SELECTION,
    SET_SEARCH_QUERY,
    SET_SORT_BY,
    SET_SORT_DIRECTION,
} = API_KEYS_ACTIONS;

declare interface ResetPagination {
    resetPageIndex(): void;
}

@Component({
    components: {
        NoApiKeysArea,
        VList,
        VHeader,
        ApiKeysItem,
        VButton,
        ApiKeysCreationPopup,
        ApiKeysCopyPopup,
        VPagination,
        SortingHeader,
        EmptySearchResultIcon,
    },
})
export default class ApiKeysArea extends Vue {
    private FIRST_PAGE = 1;
    /**
     * Indicates if delete confirmation state should appear.
     */
    private isDeleteClicked: boolean = false;
    /**
     * Indicates if api key name input state should appear.
     */
    private isNewApiKeyPopupShown: boolean = false;
    /**
     * Indicates if copy api key state should appear.
     * Should only appear once
     */
    private isCopyApiKeyPopupShown: boolean = false;
    private apiKeySecret: string = '';

    public $refs!: {
        pagination: HTMLElement & ResetPagination;
    };

    /**
     * Lifecycle hook after initial render where list of existing api keys is fetched.
     */
    public async mounted(): Promise<void> {
        await this.$store.dispatch(FETCH, 1);
        this.$segment.track(SegmentEvent.API_KEYS_VIEWED, {
            project_id: this.$store.getters.selectedProject.id,
            api_keys_count: this.selectedAPIKeysCount,
        });
    }

    /**
     * Lifecycle hook before component destruction.
     * Clears existing api keys selection and search.
     */
    public async beforeDestroy(): Promise<void> {
        this.onClearSelection();
        await this.$store.dispatch(SET_SEARCH_QUERY, '');
    }

    /**
     * toggles api key selection.
     * @param apiKey
     */
    public async toggleSelection(apiKey: ApiKey): Promise<void> {
        await this.$store.dispatch(TOGGLE_SELECTION, apiKey);
    }

    public onCreateApiKeyClick(): void {
        this.isNewApiKeyPopupShown = true;
    }

    public onFirstDeleteClick(): void {
        this.isDeleteClicked = true;
    }

    public onClearSelection(): void {
        this.$store.dispatch(CLEAR_SELECTION);
        this.isDeleteClicked = false;
    }

    public closeNewApiKeyPopup(): void {
        this.isNewApiKeyPopupShown = false;
    }

    public showCopyApiKeyPopup(secret: string): void {
        this.isCopyApiKeyPopupShown = true;
        this.apiKeySecret = secret;
    }

    public closeCopyNewApiKeyPopup(): void {
        this.isCopyApiKeyPopupShown = false;
    }

    /**
     * Deletes selected api keys, fetches updated list and changes area state to default.
     */
    public async onDelete(): Promise<void> {
        try {
            await this.$store.dispatch(DELETE);
            await this.$notify.success(`API keys deleted successfully`);
            this.$segment.track(SegmentEvent.API_KEY_DELETED, {
                project_id: this.$store.getters.selectedProject.id,
            });
        } catch (error) {
            await this.$notify.error(error.message);
        }

        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        this.isDeleteClicked = false;

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }

    public get itemComponent() {
        return ApiKeysItem;
    }

    /**
     * Returns api keys from store.
     */
    public get apiKeyList(): ApiKey[] {
        return this.$store.getters.apiKeys;
    }

    /**
     * Returns api keys pages count from store.
     */
    public get totalPageCount(): number {
        return this.$store.state.apiKeysModule.page.pageCount;
    }

    /**
     * Returns api keys label depends on api keys count.
     */
    public get apiKeyCountTitle(): string {
        return this.selectedAPIKeysCount === 1 ? 'api key' : 'api keys';
    }

    /**
     * Indicates if no api keys in store.
     */
    public get isEmpty(): boolean {
        return this.$store.getters.apiKeys.length === 0;
    }

    public get hasSearchQuery(): boolean {
        return this.$store.state.apiKeysModule.cursor.search;
    }

    public get selectedAPIKeysCount(): number {
        return this.$store.state.apiKeysModule.selectedApiKeysIds.length;
    }

    public get headerState(): number {
        return this.selectedAPIKeysCount > 0 ? HeaderState.ON_SELECT : HeaderState.DEFAULT;
    }

    public get isHeaderShown(): boolean {
        return !this.isEmpty || this.hasSearchQuery;
    }

    public get isDefaultHeaderState(): boolean {
        return this.headerState === HeaderState.DEFAULT;
    }

    public get areApiKeysSelected(): boolean {
        return this.headerState === HeaderState.ON_SELECT && !this.isDeleteClicked;
    }

    public get areSelectedApiKeysBeingDeleted(): boolean {
        return this.headerState === HeaderState.ON_SELECT && this.isDeleteClicked;
    }

    public get isEmptySearchResultShown(): boolean {
        return this.isEmpty && this.hasSearchQuery;
    }

    public get isEmptyStateShown(): boolean {
        return this.isEmpty && !this.isNewApiKeyPopupShown && !this.hasSearchQuery;
    }

    /**
     * Fetches api keys page by clicked index.
     * @param index
     */
    public async onPageClick(index: number): Promise<void> {
        try {
            await this.$store.dispatch(FETCH, index);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }
    }

    /**
     * Used for sorting.
     * @param sortBy
     * @param sortDirection
     */
    public async onHeaderSectionClickCallback(sortBy: ApiKeyOrderBy, sortDirection: SortDirection): Promise<void> {
        await this.$store.dispatch(SET_SORT_BY, sortBy);
        await this.$store.dispatch(SET_SORT_DIRECTION, sortDirection);
        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }

    /**
     * Sets api keys search query and then fetches depends on it.
     * @param query
     */
    public async onSearchQueryCallback(query: string): Promise<void> {
        await this.$store.dispatch(SET_SEARCH_QUERY, query);
        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }
}
</script>

<style scoped lang="scss">
    .api-keys-area {
        position: relative;
        padding: 40px 65px 55px 65px;
        font-family: 'font_regular', sans-serif;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 32px;
            line-height: 39px;
            color: #263549;
            margin: 0;
        }

        .api-keys-header {
            width: 100%;
            position: relative;

            .blur-content {
                position: absolute;
                top: 100%;
                left: 0;
                background-color: #f5f6fa;
                width: 100%;
                height: 70vh;
                z-index: 100;
                opacity: 0.3;
            }

            .blur-search {
                position: absolute;
                bottom: 0;
                right: 0;
                width: 540px;
                height: 56px;
                z-index: 100;
                opacity: 0.3;
                background-color: #f5f6fa;
            }
        }

        .api-keys-items {
            position: relative;

            &__content {
                display: flex;
                flex-direction: column;
                width: 100%;
                justify-content: flex-start;
            }
        }
    }

    .empty-search-result-area {
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 32px;
            line-height: 39px;
            margin-top: 104px;
        }

        &__image {
            margin-top: 40px;
        }
    }

    .pagination-area {
        margin-left: -25px;
        padding-bottom: 15px;
    }

    .header-default-state,
    .header-selected-api-keys {
        display: flex;
        align-items: center;

        &__info-text {
            margin-left: 25px;
            line-height: 48px;
        }
    }

    .header-selected-api-keys {

        .deletion {
            margin-right: 12px;
        }
    }

    .header-after-delete-click {
        display: flex;
        flex-direction: column;
        margin-top: 2px;

        &__confirmation-label {
            font-family: 'font_medium', sans-serif;
            font-size: 14px;
            line-height: 28px;
        }

        &__button-area {
            display: flex;
            margin-top: 4px;

            .button {
                margin-top: 2px;
            }

            .deletion {
                margin: 3px 12px 0 0;
            }
        }
    }

    .container.deletion {
        background-color: #ff4f4d;

        &.label {
            color: #fff;
        }

        &:hover {
            background-color: #de3e3d;
            box-shadow: none;
        }
    }

    .collapsed {
        margin-top: 0 !important;
        padding-top: 0 !important;
    }

    ::-webkit-scrollbar,
    ::-webkit-scrollbar-track,
    ::-webkit-scrollbar-thumb {
        width: 0;
    }

    @media screen and (max-width: 1024px) {

        .api-keys-area {
            padding: 40px 40px 55px 40px;
        }
    }
</style>
