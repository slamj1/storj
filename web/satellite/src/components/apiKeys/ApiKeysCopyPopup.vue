// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="save-api-popup" v-if="isPopupShown">
        <h2 class="save-api-popup__title">Save Your Secret API Key! It Will Appear Only Once.</h2>
        <div class="save-api-popup__copy-area">
            <div class="save-api-popup__copy-area__key-area">
                <p class="save-api-popup__copy-area__key-area__key">{{ apiKeySecret }}</p>
            </div>
            <p class="save-api-popup__copy-area__copy-button" @click="onCopyClick">Copy</p>
        </div>
        <div class="save-api-popup__next-step-area">
            <span class="save-api-popup__next-step-area__label">Next Step:</span>
            <a
                class="save-api-popup__next-step-area__link"
                href="https://documentation.tardigrade.io/getting-started/uploading-your-first-object/set-up-uplink-cli"
                target="_blank"
                @click.self.stop="segmentTrack"
            >
                Set Up Uplink CLI
            </a>
            <VButton
                label="Done"
                width="156px"
                height="40px"
                :on-press="onCloseClick"
            />
        </div>
        <div class="blur-content"></div>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import HeaderlessInput from '@/components/common/HeaderlessInput.vue';
import VButton from '@/components/common/VButton.vue';

import { SegmentEvent } from '@/utils/constants/analyticsEventNames';

@Component({
    components: {
        VButton,
        HeaderlessInput,
    },
})
export default class ApiKeysCopyPopup extends Vue {
    /**
     * Indicates if component should be rendered.
     */
    @Prop({default: false})
    private readonly isPopupShown: boolean;
    @Prop({default: ''})
    private readonly apiKeySecret: string;

    /**
     * Closes popup.
     */
    public onCloseClick(): void {
        this.$emit('closePopup');
    }

    /**
     * Copies api key secret to buffer.
     */
    public onCopyClick(): void {
        this.$copyText(this.apiKeySecret);
        this.$notify.success('Key saved to clipboard');
    }

    /**
     * Tracks if user checked uplink CLI docs.
     */
    public segmentTrack(): void {
        this.$segment.track(SegmentEvent.CLI_DOCS_VIEWED, {
            email: this.$store.getters.user.email,
        });
    }
}
</script>

<style scoped lang="scss">
    .save-api-popup {
        padding: 32px 40px;
        background-color: #fff;
        border-radius: 24px;
        margin-top: 29px;
        max-width: 94.8%;
        height: auto;
        position: relative;
        font-family: 'font_regular', sans-serif;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 24px;
            line-height: 29px;
            margin-bottom: 26px;
        }

        &__copy-area {
            display: flex;
            align-items: center;
            justify-content: space-between;
            background-color: #f5f6fa;
            padding: 29px 32px 29px 24px;
            border-radius: 12px;
            position: relative;
            margin-bottom: 20px;

            &__key-area {

                &__key {
                    margin: 0;
                    font-size: 16px;
                    line-height: 21px;
                    word-break: break-all;
                }
            }

            &__copy-button {
                padding: 11px 22px;
                margin: 0 0 0 20px;
                background: #fff;
                border-radius: 6px;
                font-size: 15px;
                cursor: pointer;
                color: #2683ff;

                &:hover {
                    color: #fff;
                    background-color: #2683ff;
                }
            }
        }

        &__next-step-area {
            display: flex;
            justify-content: flex-end;
            align-items: center;
            width: 100%;

            &__label {
                font-size: 15px;
                line-height: 49px;
                letter-spacing: -0.100741px;
                color: #a0a0a0;
                margin-right: 15px;
            }

            &__link {
                display: flex;
                align-items: center;
                justify-content: center;
                border: 1px solid #2683ff;
                border-radius: 6px;
                width: 154px;
                height: 38px;
                font-size: 15px;
                line-height: 22px;
                color: #2683ff;
                margin-right: 15px;

                &:hover {
                    color: #fff;
                    background-color: #2683ff;
                }
            }
        }

        .blur-content {
            position: absolute;
            top: 100%;
            left: 0;
            background-color: #f5f6fa;
            width: 100%;
            height: 70.5vh;
            z-index: 100;
            opacity: 0.3;
        }
    }
</style>
