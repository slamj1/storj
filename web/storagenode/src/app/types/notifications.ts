// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { NotificationIcon } from '@/app/utils/notificationIcons';

/**
 * Describes notification entity.
 */
export class Notification {
    public icon: NotificationIcon;

    public constructor(
        public id: string = '',
        public senderId: string = '',
        public type: NotificationTypes = NotificationTypes.Custom,
        public title: string = '',
        public message: string = '',
        public isRead: boolean = false,
        public createdAt: Date = new Date(),
    ) {
        this.setIcon();
    }

    /**
     * dateLabels formats createdAt into more informative strings.
     */
    public get dateLabel(): string {
        const differenceInSeconds = (Math.trunc(new Date().getTime()) - Math.trunc(new Date(this.createdAt).getTime())) / 1000;

        switch (true) {
            case differenceInSeconds < 60:
                return 'Just now';
            case differenceInSeconds < 3600:
                return `${(differenceInSeconds / 60).toFixed(0)} minute(s) ago`;
            case differenceInSeconds < 86400:
                return `${(differenceInSeconds / 3600).toFixed(0)} hour(s) ago`;
            case differenceInSeconds < 86400 * 2:
                return `Yesterday`;
            default:
                return this.createdAt.toDateString();
        }
    }

    /**
     * markAsRead mark notification as read on UI.
     */
    public markAsRead(): void {
        this.isRead = true;
    }

    /**
     * setIcon selects notification icon depends on type.
     */
    private setIcon(): void {
        switch (this.type) {
            case NotificationTypes.AuditCheckFailure:
                this.icon = NotificationIcon.FAIL;
                break;
            case NotificationTypes.UptimeCheckFailure:
                this.icon = NotificationIcon.DISQUALIFIED;
                break;
            case NotificationTypes.Disqualification:
                this.icon = NotificationIcon.SOFTWARE_UPDATE;
                break;
            case NotificationTypes.Suspension:
                this.icon = NotificationIcon.SUSPENDED;
                break;
            default:
                this.icon = NotificationIcon.INFO;
        }
    }
}

/**
 * Describes all current notifications types.
 */
export enum NotificationTypes {
    Custom = 0,
    AuditCheckFailure = 1,
    UptimeCheckFailure = 2,
    Disqualification = 3,
    Suspension = 4,
}

/**
 * Describes page offset for pagination.
 */
export class NotificationsCursor {
    public constructor(
        public page: number = 0,
        public limit: number = 7,
    ) { }
}

/**
 * Holds all notifications module state.
 */
export class NotificationsState {
    public latestNotifications: Notification[] = [];

    public constructor(
        public notifications: Notification[] = [],
        public pageCount: number = 0,
        public unreadCount: number = 0,
    ) { }
}

/**
 * Exposes all notifications-related functionality.
 */
export interface NotificationsApi {
    /**
     * Fetches notifications.
     * @throws Error
     */
    get(cursor: NotificationsCursor): Promise<NotificationsState>;

    /**
     * Marks single notification as read.
     * @throws Error
     */
    read(id: string): Promise<void>;

    /**
     * Marks all notification as read.
     * @throws Error
     */
    readAll(): Promise<void>;
}
