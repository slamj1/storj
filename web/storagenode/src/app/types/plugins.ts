// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { Telemetry } from '@/app/telemetry/telemetry';

declare module 'vue/types/vue' {
    interface Vue {
        $telemetry: Telemetry;
    }
}
