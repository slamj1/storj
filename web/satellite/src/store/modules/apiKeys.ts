// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { ApiKey, ApiKeyCursor, ApiKeyOrderBy, ApiKeysApi, ApiKeysPage } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';

export const API_KEYS_MUTATIONS = {
    SET_PAGE: 'setApiKeys',
    TOGGLE_SELECTION: 'toggleApiKeysSelection',
    CLEAR_SELECTION: 'clearApiKeysSelection',
    CLEAR: 'clearApiKeys',
    CHANGE_SORT_ORDER: 'changeApiKeysSortOrder',
    CHANGE_SORT_ORDER_DIRECTION: 'changeApiKeysSortOrderDirection',
    SET_SEARCH_QUERY: 'setApiKeysSearchQuery',
    SET_PAGE_NUMBER: 'setApiKeysPage',
};

const {
    SET_PAGE,
    TOGGLE_SELECTION,
    CLEAR_SELECTION,
    CLEAR,
    CHANGE_SORT_ORDER,
    CHANGE_SORT_ORDER_DIRECTION,
    SET_SEARCH_QUERY,
    SET_PAGE_NUMBER,
} = API_KEYS_MUTATIONS;

export class ApiKeysState {
    public cursor: ApiKeyCursor = new ApiKeyCursor();
    public page: ApiKeysPage = new ApiKeysPage();
    public selectedApiKeysIds: string[] = [];
}

/**
 * creates apiKeys module with all dependencies
 *
 * @param api - apiKeys api
 */
export function makeApiKeysModule(api: ApiKeysApi): StoreModule<ApiKeysState> {
    return {
        state: new ApiKeysState(),
        mutations: {
            [SET_PAGE](state: ApiKeysState, page: ApiKeysPage) {
                state.page = page;
                state.page.apiKeys = state.page.apiKeys.map(apiKey => {
                    if (state.selectedApiKeysIds.includes(apiKey.id)) {
                        apiKey.isSelected = true;
                    }

                    return apiKey;
                });
            },
            [SET_PAGE_NUMBER](state: ApiKeysState, pageNumber: number) {
                state.cursor.page = pageNumber;
            },
            [SET_SEARCH_QUERY](state: ApiKeysState, search: string) {
                state.cursor.search = search;
            },
            [CHANGE_SORT_ORDER](state: ApiKeysState, order: ApiKeyOrderBy) {
                state.cursor.order = order;
            },
            [CHANGE_SORT_ORDER_DIRECTION](state: ApiKeysState, direction: SortDirection) {
                state.cursor.orderDirection = direction;
            },
            [TOGGLE_SELECTION](state: ApiKeysState, apiKey: ApiKey) {
                if (!state.selectedApiKeysIds.includes(apiKey.id)) {
                    apiKey.isSelected = true;
                    state.selectedApiKeysIds.push(apiKey.id);

                    return;
                }

                apiKey.isSelected = false;
                state.selectedApiKeysIds = state.selectedApiKeysIds.filter(apiKeyId => {
                    return apiKey.id !== apiKeyId;
                });
            },
            [CLEAR_SELECTION](state: ApiKeysState) {
                state.selectedApiKeysIds = [];
                state.page.apiKeys = state.page.apiKeys.map((apiKey: ApiKey) => {
                    apiKey.isSelected = false;

                    return apiKey;
                });
            },
            [CLEAR](state: ApiKeysState) {
                state.cursor = new ApiKeyCursor();
                state.page = new ApiKeysPage();
                state.selectedApiKeysIds = [];
            },
        },
        actions: {
            fetchApiKeys: async function ({commit, rootGetters, state}, pageNumber: number): Promise<ApiKeysPage> {
                const projectId = rootGetters.selectedProject.id;
                commit(SET_PAGE_NUMBER, pageNumber);

                const apiKeysPage: ApiKeysPage = await api.get(projectId, state.cursor);
                commit(SET_PAGE, apiKeysPage);

                return apiKeysPage;
            },
            createApiKey: async function ({commit, rootGetters}: any, name: string): Promise<ApiKey> {
                const projectId = rootGetters.selectedProject.id;

                const apiKey = await api.create(projectId, name);

                return apiKey;
            },
            deleteApiKey: async function({state, commit}: any): Promise<void> {
                await api.delete(state.selectedApiKeysIds);

                commit(CLEAR_SELECTION);
            },
            setApiKeysSearchQuery: function ({commit}, search: string) {
                commit(SET_SEARCH_QUERY, search);
            },
            setApiKeysSortingBy: function ({commit}, order: ApiKeyOrderBy) {
                commit(CHANGE_SORT_ORDER, order);
            },
            setApiKeysSortingDirection: function ({commit}, direction: SortDirection) {
                commit(CHANGE_SORT_ORDER_DIRECTION, direction);
            },
            toggleApiKeySelection: function ({commit}, apiKey: ApiKey): void {
                commit(TOGGLE_SELECTION, apiKey);
            },
            clearApiKeySelection: function ({commit}): void {
                commit(CLEAR_SELECTION);
            },
            clearApiKeys: function ({commit}): void {
                commit(CLEAR);
                commit(CLEAR_SELECTION);
            },
        },
        getters: {
            selectedApiKeys: (state: ApiKeysState) => state.page.apiKeys.filter((key: ApiKey) => key.isSelected),
            apiKeys: function (state: ApiKeysState): ApiKey[] {
                return state.page.apiKeys;
            },
        },
    };
}
