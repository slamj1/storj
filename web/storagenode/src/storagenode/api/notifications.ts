// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { Notification, NotificationsApi, NotificationsCursor, NotificationsState } from '@/app/types/notifications';
import { HttpClient } from '@/storagenode/utils/httpClient';

/**
 * NotificationsHttpApi is a http implementation of Notifications API.
 * Exposes all notifications-related functionality
 */
export class NotificationsHttpApi implements NotificationsApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/notifications';

    /**
     * Fetch notifications.
     *
     * @returns notifications state
     * @throws Error
     */
    public async get(cursor: NotificationsCursor): Promise<NotificationsState> {
        const path = `${this.ROOT_PATH}/list?page=${cursor.page}&limit=${cursor.limit}`;
        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get notifications');
        }

        const notificationResponse = await response.json();
        let notifications: Notification[] = [];
        let pageCount: number = 0;
        let unreadCount: number = 0;

        if (notificationResponse) {
            notifications = notificationResponse.page.notifications.map(item =>
                new Notification(
                    item.id,
                    item.senderId,
                    item.type,
                    item.title,
                    item.message,
                    !!item.readAt,
                    new Date(item.createdAt),
                ),
            );

            pageCount = notificationResponse.page.pageCount;
            unreadCount = notificationResponse.unreadCount;
        }

        return new NotificationsState(notifications, pageCount, unreadCount);
    }

    /**
     * Marks single notification as read.
     * @throws Error
     */
    public async read(id: string): Promise<void> {
        const path = `${this.ROOT_PATH}/${id}/read`;
        const response = await this.client.post(path, null);

        if (response.ok) {
            return;
        }

        throw new Error('can not mark notification as read');
    }

    /**
     * Marks all notifications as read.
     * @throws Error
     */
    public async readAll(): Promise<void> {
        const path = `${this.ROOT_PATH}/readall`;
        const response = await this.client.post(path, null);

        if (response.ok) {
            return;
        }

        throw new Error('can not mark all notifications as read');
    }
}
