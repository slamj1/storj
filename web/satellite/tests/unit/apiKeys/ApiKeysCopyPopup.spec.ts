// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import VueClipboard from 'vue-clipboard2';
import Vuex from 'vuex';

import ApiKeysCopyPopup from '@/components/apiKeys/ApiKeysCopyPopup.vue';

import { ApiKeysApiGql } from '@/api/apiKeys';
import { makeApiKeysModule } from '@/store/modules/apiKeys';
import { makeNotificationsModule } from '@/store/modules/notifications';
import { NotificatorPlugin } from '@/utils/plugins/notificator';
import { createLocalVue, mount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueClipboard);
const notificationPlugin = new NotificatorPlugin();
localVue.use(notificationPlugin);
const apiKeysApi = new ApiKeysApiGql();
const apiKeysModule = makeApiKeysModule(apiKeysApi);
const notificationsModule = makeNotificationsModule();
const testKey = 'test';

const store = new Vuex.Store({ modules: { notificationsModule, apiKeysModule }});

describe('ApiKeysCopyPopup', () => {
    it('renders correctly', async (): Promise<void> => {
        const wrapper = mount(ApiKeysCopyPopup, {
            store,
            localVue,
            propsData: {
                isPopupShown: true,
                apiKeySecret: testKey,
            },
        });

        expect(wrapper).toMatchSnapshot();
        await expect(wrapper.find('.save-api-popup__copy-area__key-area__key').text()).toBe(testKey);
    });

    it('function onCloseClick works correctly', async (): Promise<void> => {
        const wrapper = mount(ApiKeysCopyPopup, {
            store,
            localVue,
        });

        await wrapper.vm.onCloseClick();

        expect(wrapper.emitted()).toEqual({'closePopup': [[]]});
    });
});
