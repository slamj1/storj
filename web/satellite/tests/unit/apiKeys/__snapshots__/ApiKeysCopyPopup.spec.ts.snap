// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiKeysCopyPopup renders correctly 1`] = `
<div class="save-api-popup">
  <h2 class="save-api-popup__title">Save Your Secret API Key! It Will Appear Only Once.</h2>
  <div class="save-api-popup__copy-area">
    <div class="save-api-popup__copy-area__key-area">
      <p class="save-api-popup__copy-area__key-area__key">test</p>
    </div>
    <p class="save-api-popup__copy-area__copy-button">Copy</p>
  </div>
  <div class="save-api-popup__next-step-area"><span class="save-api-popup__next-step-area__label">Next Step:</span> <a href="https://documentation.tardigrade.io/getting-started/uploading-your-first-object/set-up-uplink-cli" target="_blank" class="save-api-popup__next-step-area__link">
      Set Up Uplink CLI
    </a>
    <div class="container" style="width: 156px; height: 40px;"><span class="label">Done</span></div>
  </div>
  <div class="blur-content"></div>
</div>
`;
